### PR TITLE
Plugin initial implementation connect ExpressGateway/express-gateway#392

### PR DESCRIPTION
**Pull request by XVincentX**
This PR implements a rewrite plugin based on RegExp and Express Paths.

It contains 2 `conditions` and one `policy` based on multiple discussion we had during the last days.

The implementation has been partially influenced by `http-rewrite-middleware` - kudos to these guys.

- [x] Tests

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (DrMegavolt) [inline .vscode/launch.json]: can you add  gitignore with standard items like .idea , .vscode etc

File: README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+policies:
+  -
+    rewrite:
+      -
+        condition:
+          name: pathmatch
+          match: /tina/:code
+        action:
+          rewrite: /status/:code
+          redirect: permanent
```
Comment (DrMegavolt) [inline README.md]: Why not have 301,302 , 307 etc here ?

File: manifest.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            toUrl = req.url.replace(req.egContext.matchedCondition.plainRegEx, actionParams.rewrite);
+          else
+            toUrl = compiledExp(req.egContext.matchedCondition);
+
+          if (!actionParams.redirect) {
+            req.url = toUrl;
+            return next();
+          }
+
+          res.statusCode = actionParams.redirect === 'permanent' ? 301 : 302;
```
Comment (DrMegavolt) [inline manifest.js]: see comment above, 301,302 are not the only options here

File: manifest.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+          res.setHeader('Location', toUrl);
+          res.end();
```
Comment (DrMegavolt) [inline manifest.js]: why not use res.redirect(301,toUrl); ?

File: test/server.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+const path = require('path');
+const gateway = require('express-gateway');
+
+gateway()
+  .load(path.join(__dirname, 'config'))
+  .run();
```
Comment (DrMegavolt) [inline test/server.js]: I don't see any test here except it runs. 
Minimal variant is to run real server + fake backend server and ensure that fake backend is recieving correct url
same for redirect

File: test/config/gateway.config.yml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  - name: basic
+    apiEndpoints:
+      - api
+    policies:
+      -
+        rewrite:
+          -
+            condition:
+              name: pathmatch
+              match: /gigino/:code
```
Comment (DrMegavolt) [inline test/config/gateway.config.yml]: Could you please check how variant with multi-segment match is working 
match /api/:route*
rewrite /api/v1/:route

incoming url  /api/users/nick
proxied to /api/v1/users/nick

File: .vscode/launch.json (hunk 1 above)
Comment (XVincentX) [reply .vscode/launch.json]: Definitely!

File: manifest.js (hunk 4 above)
Comment (XVincentX) [reply manifest.js]: Good point, I think it's handling more cases than I could imagine. I'll fix it.

File: test/server.js (hunk 5 above)
Comment (XVincentX) [reply test/server.js]: Definitely - as I wrote in the pull request text, I'm figuring out how to test but more importantly - I'm missing a way to shut down the gateway programmatically. I can see from the gateway repo you guys have an utility to fork and kill the process and I'm not entirely sure whether I should just copy paste that code here or eventually adding a shutdown function to the gateway. What's your opinion?

File: test/config/gateway.config.yml (hunk 6 above)
Comment (XVincentX) [reply test/config/gateway.config.yml]: 👍 I'll code all these as test cases.

File: README.md (hunk 2 above)
Comment (XVincentX) [reply README.md]: Good point - I honestly ignored all the other status codes as they felt "unused" to me. I'll have a quick look to the spec again and modify the code accordingly.

**Comment by XVincentX** [comment]
@DrMegavolt I've addressed all your comments as well added several tests to make sure the gateway is interacting correctly with the plugin.

Let me know if you want me to go further and add unit tests for the functions as well - although I do not see any great advantage here as most of the work is done by the internal dependency.

File: test/server.js (hunk 5 above)
Comment (XVincentX) [reply test/server.js]: Addressed!

File: plugin.test.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
+const axios = require('axios').default;
+const path = require('path');
+const gateway = require('express-gateway');
+const express = require('express');
+
+let Application = undefined;
+
+beforeAll((done) => {
+  axios.defaults.baseURL = 'http://localhost:8080/';
```
Comment (stanislaw-glogowski) [inline plugin.test.js]: You should consider using `axios.create` instead of changing configuration of default instance of axios.

Comment (XVincentX) [reply plugin.test.js]: @stanislaw-glogowski Done

**Pull request merged**